### PR TITLE
Renamings, cleanup and UState.from_uctx which respects invariants

### PR DESCRIPTION
**Pull request by yannl35133**
The change for universe binder printing will make it slower, but I don't know how often it happens and if I shoud worry about it.

Overlays: https://github.com/rocq-prover/vsrocq/pull/1218

### CODE REVIEW

File: engine/uState.mli [hunk 1]
```diff
@@ -32,20 +32,17 @@ type t
...
+val from_env : Environ.env -> t
+(** Main entry point at the beginning of a declaration. *)
 
-val make_with_initial_binders : qualities:QGraph.t -> UGraph.t -> lident list -> t
-[@@ocaml.deprecated "(8.13) Use from_env"]
-
-val from_env : ?binders:lident list -> Environ.env -> t
-(** Main entry point at the beginning of a declaration declaring the
-    binding names as rigid universes. *)
+val from_uctx : Environ.env -> UVars.UContext.t -> t
```
Comment (SkySkimmer) [inline engine/uState.mli]: why not
```suggestion
val add_uctx : t -> UVars.UContext.t -> t
```
instead of combining with from_env?

File: test-suite/output/UnivBinders.out [hunk 2]
```diff
@@ -88,10 +88,10 @@ foo@{uu u v} =
 Type@{u} -> Type@{v} -> Type@{uu}
      : Type@{max(uu+1,u+1,v+1)}
 (* uu u v |=  *)
-foo@{u u IMPORTANT} =
-Type@{u} -> Type@{IMPORTANT} -> Type@{u}
-     : Type@{max(u+1,u+1,IMPORTANT+1)}
-(* u u IMPORTANT |=  *)
+foo@{u u0 IMPORTANT} =
```
Comment (SkySkimmer) [inline test-suite/output/UnivBinders.out]: bug?

Comment (yannl35133) [reply test-suite/output/UnivBinders.out]: No, creating a valid ustate means that all levels have distinct names. So they get freshened.
I think this is enough of an edge case not to be reported though.

File: vernac/himsg.ml [hunk 3]
```diff
@@ -1305,14 +1305,12 @@ let explain_not_match_error = function
...
     let open UVars in
     let pr_auctx auctx =
-      let sigma = Evd.from_ctx
-          (UState.of_names
-             (Printer.universe_binders_with_opt_names auctx None))
-      in
-      let uctx = AbstractContext.repr auctx in
+      let uctx = UVars.AbstractContext.repr auctx in
+      let uctx = Printer.fill_names uctx in
+      let sigma = Evd.from_uctx env uctx in
```
Comment (SkySkimmer) [inline vernac/himsg.ml]: it seems like this pattern (`from_uctx env (fill_names (repr auctx))`) is the only way fill_names gets used, maybe we should expose it instead of fill_names?

File: engine/uState.mli (hunk 1 above)
Comment (yannl35133) [reply engine/uState.mli]: It would be fine code-wise, but is there a reason to require the intermediate step or to need to add two uctxs ?

File: test-suite/output/UnivBinders.out (hunk 2 above)
Comment (SkySkimmer) [reply test-suite/output/UnivBinders.out]: being able to have multiple levels with the same name is deliberate

Comment (SkySkimmer) [reply test-suite/output/UnivBinders.out]: (in the test source: `(* Can use a name for multiple universes *)`)

File: vernac/himsg.ml (hunk 3 above)
Comment (yannl35133) [reply vernac/himsg.ml]: `fill_names` is rather ugly, so I don't want to put it into Evd or UContext, and I can't see a better interface. I'm open to suggestions though (including if you convince to me to actually put `fill_names` in Evd / UContext).

File: engine/uState.mli (hunk 1 above)
Comment (SkySkimmer) [reply engine/uState.mli]: this API makes ucontext look more special than it IMO deserves

File: vernac/himsg.ml (hunk 3 above)
Comment (SkySkimmer) [reply vernac/himsg.ml]: I mean have `Printer.evd_from_auctx : env -> auctx -> evar_map` instead of exposing fill_names

File: test-suite/output/UnivBinders.out (hunk 2 above)
Comment (yannl35133) [reply test-suite/output/UnivBinders.out]: I got mixed up with the situation in `bug_4519.v` where this change is a positive for me.

Comment (SkySkimmer) [reply test-suite/output/UnivBinders.out]: that one is fine to change (and maybe turn it into an output test)

File: engine/uState.mli (hunk 1 above)
Comment (yannl35133) [reply engine/uState.mli]: Would you rather it be `AbstractContext.t` or `universes`?

Comment (SkySkimmer) [reply engine/uState.mli]: either's fine

File: vernac/himsg.ml (hunk 3 above)
Comment (yannl35133) [reply vernac/himsg.ml]: I would rather introduce a UnivPrinter structure only featuring the names component of the ustate and remove the need for the `env` argument, but that would become outside the scope of this PR.

File: test-suite/output/UnivBinders.out (hunk 2 above)
Comment (yannl35133) [reply test-suite/output/UnivBinders.out]: I added the test to UnivBinders.v instead.

File: vernac/himsg.ml (hunk 3 above)
Comment (SkySkimmer) [reply vernac/himsg.ml]: Is that even possible? detyping needs a full evar map

Comment (yannl35133) [reply vernac/himsg.ml]: Ah, I got too focused on how universes and constraints got printed, I forgot about that part.

Comment (yannl35133) [reply vernac/himsg.ml]: > I mean have `Printer.evd_from_auctx : env -> auctx -> evar_map` instead of exposing fill_names

This could be done in a later PR, but doesn't need to happen here.

Comment (SkySkimmer) [reply vernac/himsg.ml]: it is better to do it in this PR rather than exposing fill_names

Comment (yannl35133) [reply vernac/himsg.ml]: Is `fill_names` really worse than `universe_binders_with_opt_names`?

**Review by SkySkimmer — APPROVED** [review]

**Comment by SkySkimmer** [comment]
@coqbot merge now
